src/pipeline.py

```diff
@@ -160,29 +161,44 @@ def generate_grid(locations):
 
 def get_location_descriptors(mixvpr_agg, cpip_model, locations):
     location_encoder = cpip_model.location_projection
-    with torch.no_grad():
-        encoded_locations = location_encoder(locations)  # Encode all location vectors
-        # let num_points = ((2*CFG.grid_extent+1)**2 - 1) * CFG.num_rotation_steps,
-        # encoded_locations shape: (b, num_points, CFG.contrastive_dimension)
+    cnn_model = cpip_model.cnn
+    dotproduct_model = cpip_model.dotproduct
+
+    combined_data_batches = []
+    batch_size = 25
 
-        # current contrastive_dimension = 256
-        # Reshape to [1, sqrt(256), sqrt(256)] for each vector
-        b, num_points, _ = encoded_locations.shape
-        dimension_size = int(CFG.contrastive_dimension**0.5)
-        reshaped_locations = encoded_locations.view(b * num_points, 1, dimension_size, dimension_size)
+    with torch.no_grad():
+        total_batches = math.ceil(locations.size(0) / batch_size)
+        
+        for i in range(total_batches):
+            start_idx = i * batch_size
+            end_idx = min((i + 1) * batch_size, locations.size(0))
+            
+            batch_locations = locations[start_idx:end_idx] # Shape: (batchsize, numpoints, 3) = [25, 480, 3]
+            encoded_locations = location_encoder(batch_locations)  # Encode all location vectors
+            # Shape: (batchsize, numpoints, CFG.contrastive_dimension) = [25, 480, 256]
+            # let num_points = ((2*CFG.grid_extent+1)**2 - 1) * CFG.num_rotation_steps,
+            # encoded_locations shape: (b, num_points, CFG.contrastive_dimension)
 
-        # Create and apply CNN model
-        cnn_model = ShallowConvNet(input_c=1, input_h=dimension_size, input_w=dimension_size, output_c=1024, output_h=20, output_w=20).to(CFG.device)
-        # TODO: cnn weights
-        cnn_model.eval()
-        descriptors = cnn_model(reshaped_locations)
+            # current contrastive_dimension = 256
+            # Reshape to [1, sqrt(256), sqrt(256)] for each vector
+            b, num_points, _ = encoded_locations.shape
+            dimension_size = int(CFG.contrastive_dimension**0.5)
+            reshaped_locations = encoded_locations.view(b * num_points, 1, dimension_size, dimension_size)# [24000, 1, 16, 16] 
+            
+            descriptors = cnn_model(reshaped_locations) 
+            # descriptors: [12000, 1024, 20, 20]
+            pdb.set_trace()
```

Comment:
let's remove all pdb related lines of code
i suggest using `python -m debugpy --listen 5678 --wait-for-client your_python_file.py args` to use the debugger so you don't need to put debugger related code in the files. this works with the vscode remote attach debugger ui.